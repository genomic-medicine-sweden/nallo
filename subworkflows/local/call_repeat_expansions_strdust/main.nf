include { STRDUST          } from '../../../modules/nf-core/strdust/'
include { ADD_FOUND_IN_TAG } from '../../../modules/local/add_found_in_tag/main'
include { BCFTOOLS_MERGE   } from '../../../modules/nf-core/bcftools/merge/'

workflow CALL_REPEAT_EXPANSIONS_STRDUST {
    take:
    ch_bam_bai // channel: [mandatory] [ val(meta), path(bam), path(bai) ]
    ch_fasta   // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_fai     // channel: [mandatory] [ val(meta), path(fai) ]
    ch_bed     // channel: [mandatory] [ val(meta), path(bed) ]

    main:
    ch_versions = Channel.empty()

    STRDUST(
        ch_bam_bai,
        ch_fasta,
        ch_fai,
        ch_bed,
    )
    ch_versions.mix(STRDUST.out.versions)

    ADD_FOUND_IN_TAG(
        STRDUST.out.vcf.join(STRDUST.out.tbi),
        "STRdust",
    )

    ADD_FOUND_IN_TAG.out.vcf
        .join(ADD_FOUND_IN_TAG.out.tbi, failOnDuplicate: true, failOnMismatch: true)
        .map { meta, vcf, tbi -> [[id: meta.family_id], vcf, tbi] }
        .groupTuple()
        .set { ch_bcftools_merge_in }

    BCFTOOLS_MERGE(
        ch_bcftools_merge_in,
        [[], []],
        [[], []],
        [[], []],
    )
    ch_versions = ch_versions.mix(BCFTOOLS_MERGE.out.versions)

    emit:
    sample_vcf = STRDUST.out.vcf // channel: [ val(meta), path(vcf) ]
    sample_tbi = STRDUST.out.tbi // channel: [ val(meta), path(tbi) ]
    family_vcf = BCFTOOLS_MERGE.out.vcf // channel: [ val(meta), path(vcf) ]
    family_tbi = BCFTOOLS_MERGE.out.index // channel: [ val(meta), path(tbi) ]
    versions   = ch_versions // channel: [ versions.yml ]
}
