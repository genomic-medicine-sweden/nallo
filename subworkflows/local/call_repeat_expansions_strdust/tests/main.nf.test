nextflow_workflow {

    name "Test Subworkflow CALL_REPEAT_EXPANSIONS_STRDUST"
    script "../main.nf"
    workflow "CALL_REPEAT_EXPANSIONS_STRDUST"

    setup {
            run("GUNZIP") {
                script "../../../../modules/nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test' ],
                        file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                    ]
                    """
                }

            }

            run("SAMTOOLS_FAIDX") {
                script "../../../../modules/nf-core/samtools/faidx/main.nf"
                process {
                    """
                    input[0] = GUNZIP.out.gunzip
                    input[1] = [[],[]]
                    """
                }

            }
    }

    test("[bam, bai], fasta, fai, bed - STRdust") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false, family_id: 'family', sex: 1 ], // meta map
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.bam.bai', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = Channel.of([
                    [ id: 'pathogenic_repeats' ],
                    file(params.pipelines_testdata_base_path + 'reference/pathogenic_repeats.hg38.bed')
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.summary,
                    path(workflow.out.family_vcf.get(0).get(1)).vcf.summary,
                    workflow.out.versions,
                ).match() }
            )
        }
    }

    test("[bam, bai], fasta, fai, bed - STRdust - stub") {

        options '-stub'

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false, family_id: 'family', sex: 1 ], // meta map
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.bam.bai', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = Channel.of([
                    [ id: 'pathogenic_repeats' ],
                    file(params.pipelines_testdata_base_path + 'reference/pathogenic_repeats.hg38.bed')
                ])
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
