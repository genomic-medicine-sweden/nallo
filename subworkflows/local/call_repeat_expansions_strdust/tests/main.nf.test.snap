{
    "[bam, bai], fasta, fai, bed": {
        "content": [
            "e4981640adf7b05b75ba09d55a7185ad",
            "32e3c59f20cb0d43886a95edd73b5004",
            [
                "versions.yml:md5,3f121f951f5b25077dbdcc97a9b2d113",
                "versions.yml:md5,5e586dd4ba11708f6b7f46f58fa6dd7b"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-08T12:51:09.426007699"
    },
    "[bam, bai], fasta, fai, bed - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "family"
                        },
                        "family_repeat_expansions.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "family"
                        },
                        "family_repeat_expansions.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,3f121f951f5b25077dbdcc97a9b2d113",
                    "versions.yml:md5,5e586dd4ba11708f6b7f46f58fa6dd7b"
                ],
                "family_tbi": [
                    [
                        {
                            "id": "family"
                        },
                        "family_repeat_expansions.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "family_vcf": [
                    [
                        {
                            "id": "family"
                        },
                        "family_repeat_expansions.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "sample_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sample_vcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "family_id": "family",
                            "sex": 1
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3f121f951f5b25077dbdcc97a9b2d113",
                    "versions.yml:md5,5e586dd4ba11708f6b7f46f58fa6dd7b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-08T12:52:02.750908426"
    }
}